gcc/cp/contracts.cc

```diff
@@ -1929,6 +1932,28 @@ finish_contract_condition (cp_expr condition)
   return condition_conversion (condition);
 }
 
+/* constify access to an id from within the contract condition */
+tree
+constify_contract_access(tree decl,location_t location)
+{
+
+  /* only constifies the automatic storage variables for now.
+   * The postcondition variable is created const. Parameters need to be
+   * checked separately, and we also need to make references and *this const
+   */
+
+  if (!TREE_READONLY (decl)
+      && ((VAR_P (decl) && decl_storage_duration (decl) == dk_auto)
+          || (INDIRECT_REF_P(decl) && decl_storage_duration (TREE_OPERAND (decl, 0)) == dk_auto)
+	  || (TREE_CODE (decl) == PARM_DECL)))
+  {
+      decl = build1_loc (location, VIEW_CONVERT_EXPR,
```

Comment:
There's another bit in finish_id_expression_1 where this is done with
          tree ctype = TREE_TYPE (r);
          ctype = cp_build_qualified_type (ctype, (cp_type_quals (ctype)
                                                   | TYPE_QUAL_CONST));
          r = build1 (VIEW_CONVERT_EXPR, ctype, r);
I think we should strive to do it similarly here, and maybe refactor the two constifications into one helper function.